working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1748,65 +1746,196 @@ types of all of the case expressions.
 
 ### Variables and scope
 
-Patterns often exist to introduce new variable bindings. A "wildcard" identifier
-named `_` in a variable pattern never introduces a binding.
-The variables a patterns binds depend on what kind of pattern it is:
+Patterns often exist to bind new variables. The language must ensure that the
+variables bound by a pattern can only be used when the pattern has matched,
+which means variables bound by refutable patterns must only be in scope in code
+that can't be reached when the match fails.
+
+Also, logical-or patterns and switch case fallthrough adds some complexity.
+
+#### Pattern variable sets
+
+A *pattern variable set* specifies the set of variables declared by a pattern
+and its subpatterns when not in an assignment context. Each variable in the set
+has a unique name, a static type, and whether it is final or not. The pattern
+variable set for a pattern is:
 
-*   **Logical-or**: Does not introduce variables but may contain subpatterns
-    that do. It is a compile-time error if the two subpatterns do not introduce
-    the same variables with the same names and types.
+*   **Logical-or**: The pattern variable set of either branch. It is a
+    compile-time error if the two branches do not have equal pattern variable
+    sets. Two pattern variable sets are equal if they have the same set of names
+    and each corresponding pair of variables have the same type and finality.
 
-    **TODO: "Same name" is underspecified and also probably
-    [too conservative][2473].**
+    *Since only one branch will match and we don't know which, for the pattern
+    to have a stable set of variables with known types, the two branches must
+    define the same variables. This way, uses of the variables later will have
+    a known type and finality regardless of which branch matched.*
 
 *   **Logical-and**, **cast**, **null-check**, **null-assert**,
-    **parenthesized**, **list**, **map**, **record**, or **extractor**: These do
-    not introduce variables themselves but may contain subpatterns that do.
+    **parenthesized**, **list**, **map**, **record**, or **extractor**: The
+    union of the pattern variable sets of all of the subpatterns.
+
+    The union of a series of pattern variable sets is the union of their
+    corresponding sets of variable names. Each variable in the resulting set is
+    mapped to the corresponding variable's type and finality.
+
+    It is a compile-time error if any two sets being unioned have a variable
+    with the same name. *A pattern can't declare the same variable more than
+    once.*
+
+*   **Relational** or **constant**: An empty set.
```

Comment:
Nit: I would say "the empty set" here (and on line 1789) rather than "an empty set".  In math-speak, there is only one empty set.

Reply:
Done.